### PR TITLE
[diffusion] lora: strip `transformer.` prefix for diffusers PEFT Z-Im…

### PR DESCRIPTION
**Pull request by smidolt**
<head></head><h2 style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">Motivation</h2><p style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">Z-Image LoRAs saved through diffusers'<span class="Apple-converted-space"> </span><code>ZImagePipeline.save_lora_weights()</code><span class="Apple-converted-space"> </span>don't work in SGLang — every call to<span class="Apple-converted-space"> </span><code>set_lora</code><span class="Apple-converted-space"> </span>+<span class="Apple-converted-space"> </span><code>merge_lora_weights</code><span class="Apple-converted-space"> </span>succeeds, but the output is byte-identical to baseline. The<span class="Apple-converted-space"> </span><code>strength</code><span class="Apple-converted-space"> </span>parameter has no visible effect either (0.85 and 1.0 produce the same image).</p><p style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">Root cause: diffusers'<span class="Apple-converted-space"> </span><code>_pack_dict_with_prefix</code><span class="Apple-converted-space"> </span>(lora_base.py#L316) prepends<span class="Apple-converted-space"> </span><code>transformer.</code><span class="Apple-converted-space"> </span>to every key on save.<span class="Apple-converted-space"> </span><code>ZImageTransformer2DModel.named_modules()</code><span class="Apple-converted-space"> </span>here has no such prefix, so the key-matching loop in<span class="Apple-converted-space"> </span><code>_apply_lora_to_layers</code><span class="Apple-converted-space"> </span>(lora_pipeline.py#L436-L438) skips every single entry. No error, no warning surfaced to the user — just a silent no-op.</p><p style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">Qwen-Image has the same situation and it's already handled in this file via<span class="Apple-converted-space"> </span><code>_convert_qwen_image_standard</code>. This PR adds the equivalent path for Z-Image.</p><h3 style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">Repro before the PR</h3><pre style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;"><code class="language-python"># train any Z-Image LoRA with diffusers + PEFT
ZImagePipeline.save_lora_weights(
    save_dir,
    transformer_lora_layers=get_peft_model_state_dict(transformer, adapter_name="char"),
)
# serve that file via SGLang, POST /v1/set_lora + /v1/merge_lora_weights
# → output identical to no-LoRA baseline
</code></pre><h2 style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">Modifications</h2><p style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;"><code>python/sglang/multimodal_gen/runtime/pipelines_core/lora_format_adapter.py</code></p><ul style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;"><li><code>_looks_like_zimage_diffusers</code><span class="Apple-converted-space"> </span>— detects the format. Requires<span class="Apple-converted-space"> </span><code>transformer.</code><span class="Apple-converted-space"> </span>prefix,<span class="Apple-converted-space"> </span><code>.lora_A/B</code><span class="Apple-converted-space"> </span>substring,<span class="Apple-converted-space"> </span><strong>and</strong><span class="Apple-converted-space"> </span>a Z-Image specific branch (<code>noise_refiner</code><span class="Apple-converted-space"> </span>or<span class="Apple-converted-space"> </span><code>context_refiner</code>). The last condition keeps this from firing on Qwen-Image or any other diffusers LoRA that happens to have the<span class="Apple-converted-space"> </span><code>transformer.</code><span class="Apple-converted-space"> </span>prefix.</li><li><code>_convert_zimage_diffusers_standard</code><span class="Apple-converted-space"> </span>— drops the<span class="Apple-converted-space"> </span><code>transformer.</code><span class="Apple-converted-space"> </span>prefix. That's it.</li><li>Dispatched from<span class="Apple-converted-space"> </span><code>convert_lora_state_dict_by_format</code><span class="Apple-converted-space"> </span>in the STANDARD branch, right after the Qwen-Image check.</li></ul><p style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;"><code>python/sglang/multimodal_gen/test/unit/test_lora_format_adapter.py</code></p><ul style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;"><li><code>test_zimage_diffusers_peft_prefix_stripped</code><span class="Apple-converted-space"> </span>— builds a minimal state dict matching diffusers PEFT output (72 keys, covers<span class="Apple-converted-space"> </span><code>layers</code><span class="Apple-converted-space"> </span>/<span class="Apple-converted-space"> </span><code>noise_refiner</code><span class="Apple-converted-space"> </span>/<span class="Apple-converted-space"> </span><code>context_refiner</code><span class="Apple-converted-space"> </span>× attention + feed_forward), runs<span class="Apple-converted-space"> </span><code>normalize_lora_state_dict</code>, asserts the prefix is gone and tensor shapes are preserved. No HF download needed.</li></ul><h2 style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">Accuracy Tests</h2><p style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">Verified end-to-end on a rank-32 Z-Image-Turbo character LoRA, seed 42, 9 steps, CFG 0.0:</p>
  | baseline md5 | with LoRA md5
-- | -- | --
before PR, strength=0.85 | 3d1065a4 | 3d1065a4 (no effect)
before PR, strength=1.0 | 3d1065a4 | 3d1065a4 (no effect)
after PR, strength=0.5 | 3d1065a4 | e2d2f74a
after PR, strength=0.85 | 3d1065a4 | 74fc4fac
after PR, strength=1.0 | 3d1065a4 | 1a286cb0

<p style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">Before: baseline and LoRA outputs are the same file. After: each strength produces a distinct image and the expected character identity shows up.</p><h2 style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">Speed Tests and Profiling</h2><p style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">Not applicable. One extra prefix check at LoRA load time, runs once per file, not per request.</p><h2 style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">Checklist</h2><ul style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;"><li>[x] pre-commit hooks pass</li><li>[x] unit test added</li><li>[ ] docs — not needed, fix is transparent to users</li><li>[x] accuracy numbers above</li><li>[x] matches the existing<span class="Apple-converted-space"> </span><code>_convert_qwen_image_standard</code><span class="Apple-converted-space"> </span>structure</li></ul>

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/pipelines_core/lora_format_adapter.py [hunk 1]
```diff
@@ -105,6 +105,27 @@ def _looks_like_qwen_image(state_dict: Mapping[str, torch.Tensor]) -> bool:
...
+    if not keys:
+        return False
+    has_prefix = _has_prefix_key(keys, "transformer.")
+    has_lora_ab = _has_substring_key(keys, ".lora_A") or _has_substring_key(
+        keys, ".lora_B"
+    )
+    has_refiner = _has_prefix_key(
+        keys, "transformer.noise_refiner."
+    ) or _has_prefix_key(keys, "transformer.context_refiner.")
+    return has_prefix and has_lora_ab and has_refiner
```
[the remark this answers is not in the document]
Comment (smidolt) [reply python/sglang/multimodal_gen/runtime/pipelines_core/lora_format_adapter.py]: Good catch. An attention-only / FFN-only LoRA without refiner keys would
slip past. Adding `transformer.layers.` alone is a bit broad though — it
would also match ErnieImage attention-only LoRAs (same key shape after
PEFT export).

Tightening the check with Z-Image's Gated-MLP fingerprint (`feed_forward.w1/w2/w3`,
unique to `ZImageTransformer2DModel.FeedForward`) catches FFN-only LoRAs
without widening the matching surface. Pure attention-only Z-Image LoRAs
with no refiner keys remain a gap — but that's a rarely-used training
config, and I'd rather leave it explicit than risk misfiring on Ernie.

Updating.

[the remark this answers is not in the document]
Comment (smidolt) [reply python/sglang/multimodal_gen/runtime/pipelines_core/lora_format_adapter.py]: Pushed ef5135a. Detector now also fires on feed_forward.w1/w2/w3 — covers FFN-only LoRAs, doesn't widen to transformer.layers.* so Ernie attention-only still skipped. New test for the FFN-only case.

**Comment by smidolt** [comment]
@mickqian @ping1jing2 @yhyang201 — small Z-Image LoRA fix here (detector + test), CI is waiting on the run-ci label. Could one of you add it when you get a moment? Mirrors the existing Qwen-Image converter, no external fixtures.
